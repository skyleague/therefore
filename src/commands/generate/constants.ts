import { name } from '../../../package.json'

export const generatedBy = `Generated by ${name}`
export const generatedAjv = `Generated by Ajv https://ajv.js.org/guide/managing-schemas.html#standalone-validation-code`
