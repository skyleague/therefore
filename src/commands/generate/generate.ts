import { generatedAjv, generatedBy } from './constants.js'
import { expandGlobs } from './glob.js'
import { formatFile, maybeLoadPrettier } from './prettier.js'
import { resolveTypescriptSchema } from './resolver.js'
import type { FileDefinition, OutputFile, ReferenceData, ThereforeOutputType } from './types.js'

import { getExtension, replaceExtension } from '../../common/template/path.js'
import { renderTemplate } from '../../common/template/template.js'
import type { ThereforeNode } from '../../lib/cst/cst.js'
import type { ThereforeCst } from '../../lib/primitives/types.js'
import { isThereforeExport } from '../../lib/primitives/types.js'
import { toJsonSchema } from '../../lib/visitor/jsonschema/jsonschema.js'
import { prepass } from '../../lib/visitor/prepass/prepass.js'
import { toTypescriptDefinition } from '../../lib/visitor/typescript/typescript.js'

import { entriesOf, enumerate, groupBy, isDefined, range, second } from '@skyleague/axioms'
import decamelize from 'decamelize'

import fs from 'node:fs'
import path from 'node:path'
import { fileURLToPath } from 'node:url'

const __dirname = path.dirname(fileURLToPath(import.meta.url))
async function requireModule(module: string): Promise<Record<string, ThereforeNode | unknown>> {
    const relative = path.relative(__dirname, module).replace(/\\/g, '/').replace(/\.ts$/, '.js')

    const mod = (await import(relative.startsWith('.') ? relative : `./${relative}`)) as Record<string, unknown>
    return (mod.default ?? mod) as Record<string, ThereforeNode | unknown>
}

export async function loadSymbol({
    srcPath,
    basePath,
    symbol,
    sourceSymbol,
    compile,
    definitions,
    entry,
    fileSuffix,
    filePath,
    outputFileRename,
    appendSourceSymbolName: appendSourceName = false,
    rootPath = srcPath,
}: {
    srcPath: string
    basePath: string
    symbol: ThereforeCst
    sourceSymbol: string
    compile: boolean
    definitions: Record<string, FileDefinition | undefined>
    entry: string
    fileSuffix?: string | undefined
    filePath?: string | undefined
    outputFileRename: (path: string) => string
    appendSourceSymbolName?: boolean
    rootPath?: string
}) {
    const simplified = prepass(symbol)
    let targetDir = '.'
    if (simplified.type === 'custom') {
        if (simplified.value.fileSuffix !== undefined) {
            fileSuffix = simplified.value.fileSuffix
        }
        if (simplified.value.filePath !== undefined) {
            const dir = path.dirname(simplified.value.filePath)
            const basename = path.basename(simplified.value.filePath)
            filePath = basename
            targetDir = dir
        }
    }

    const { base: baseName, dir: baseDir } = path.parse(srcPath)
    const ext = getExtension(srcPath) ?? ''
    const targetBaseDir = path.relative(basePath, path.join(baseDir, targetDir))
    const targetSrcPath = `${targetBaseDir}/${(filePath ?? baseName).replace(ext, '')}${ext}`
    const targetPath = fileSuffix !== undefined ? replaceExtension(targetSrcPath, fileSuffix) : outputFileRename(targetSrcPath)

    definitions[targetPath] ??= {
        srcPath: path.relative(basePath, rootPath),
        targetPath,
        attachedFiles: [],
        symbols: [],
        dependencies: {},
    }

    const file = definitions[targetPath]!
    // check if symbol is not seen yet
    if (file.symbols.find((s) => s.uuid === simplified.uuid)) {
        return
    }
    const symbolName = simplified.name ?? sourceSymbol

    sourceSymbol = appendSourceName && simplified.name !== undefined ? `${sourceSymbol}.${simplified.name}` : symbolName
    const schemaName = decamelize(symbolName, { separator: '-' })
    const schemaFile = `./schemas/${schemaName}.schema.json`
    const compiledFile = `./schemas/${schemaName}.schema.js`

    const nodeIsCompiled = simplified.description.validator?.compile ?? compile

    const jsonschema = await toJsonSchema(simplified, nodeIsCompiled)
    const { definition, subtrees } = toTypescriptDefinition({ sourceSymbol, symbolName, schema: simplified })

    file.symbols.push(
        ...Object.values(definition.locals ?? {})
            .filter(isDefined)
            .map((local) => ({
                uuid: local.uuid,
                symbolName: local.symbolName,
                definition: local,
                typeOnly: true,
            }))
    )

    file.symbols.push({
        uuid: simplified.uuid,
        symbolName,
        definition: definition,
        schemaFile,
        compiledFile: nodeIsCompiled ? compiledFile : undefined,
        typeOnly: simplified.description.validator?.enabled !== true,
    })

    if (definition.isExported && simplified.description.validator?.enabled) {
        const cleanSchemasFolder = () => {
            const targetFolder = path.dirname(path.join(targetBaseDir, compiledFile))
            if (fs.existsSync(targetFolder)) {
                console.warn(`Cleaning ${targetFolder}`)
                fs.rmSync(targetFolder, { force: true, recursive: true })
            }
        }
        if (jsonschema.compiled) {
            file.attachedFiles.push({
                targetPath: path.join(targetBaseDir, compiledFile),
                content: `/**\n * ${generatedAjv} \n * eslint-disable\n */\n${jsonschema.code}`,
                prettify: false,
                type: 'validator',
                clean: cleanSchemasFolder,
            })
        } else {
            file.attachedFiles.push({
                targetPath: path.join(targetBaseDir, schemaFile),
                content: JSON.stringify(jsonschema.schema, null, 2),
                prettify: true,
                type: 'jsonschema',
                clean: cleanSchemasFolder,
            })
        }

        console.debug(` - found ${definition.symbolName}`)
    }
    for (const { node: subSymbol, fileSuffix: subFileSuffix, filePath: subFilePath } of subtrees) {
        await loadSymbol({
            symbol: subSymbol,
            basePath,
            sourceSymbol,
            compile,
            definitions,
            srcPath: path.join(basePath, targetPath),
            rootPath: srcPath,
            fileSuffix: subFileSuffix,
            filePath: subFilePath,
            entry,
            outputFileRename,
            appendSourceSymbolName: true,
        })
    }
}

export async function scanModule({
    entry,
    srcPath,
    basePath,
    compile,
    definitions,
    require = requireModule,
    outputFileRename,
}: {
    entry: string
    srcPath: string
    basePath: string
    compile: boolean
    definitions: Record<string, FileDefinition>
    require?: (module: string) => Promise<Record<string, ThereforeNode | unknown>>
    outputFileRename: (path: string) => string
}) {
    const module = await require(entry)

    for (const [sourceSymbol, symbolPromise] of Object.entries(module)) {
        const symbol = await symbolPromise

        if (!isThereforeExport(symbol)) {
            continue
        }

        await loadSymbol({ symbol, sourceSymbol, compile, definitions, srcPath, entry, basePath, outputFileRename })
    }
}

export async function scanFiles({
    files,
    basePath,
    compile,
    outputFileRename,
}: {
    files: string[]
    basePath: string
    compile: boolean
    outputFileRename: (path: string) => string
}): Promise<Record<string, FileDefinition>> {
    const definitions: Record<string, FileDefinition> = {}
    for (const entry of files) {
        const srcPath = path.resolve(basePath, entry)
        let isGenerated = false
        const strippedPath = srcPath.substr(0, srcPath.lastIndexOf('.'))
        for (const ext of ['.ts', '.js']) {
            const p = `${strippedPath}${ext}`
            isGenerated ||= fs.existsSync(p) && fs.readFileSync(p).includes(generatedBy)
        }
        if (isGenerated) {
            console.debug(`scanning ${entry}`)
            console.debug(` * skipping generated schema`)
            continue
        }

        console.log(`scanning ${entry}`)
        try {
            await scanModule({ entry, basePath, srcPath, compile, definitions, outputFileRename })
        } catch (e: unknown) {
            const error = e as Error
            console.debug(error.message, error.stack)
            throw error
        }
    }
    return definitions
}

export async function compileOutputFiles(
    entries: string[],
    { compile = false, cwd, outputFileRename }: { cwd: string; compile: boolean; outputFileRename: (path: string) => string }
): Promise<
    {
        targetPath: string
        template: string
        data: Record<string, ReferenceData>
        type: ThereforeOutputType
        prettify: boolean
        clean?: () => void
    }[]
> {
    const outputFiles: OutputFile[] = []
    const definitions = await scanFiles({ files: entries, basePath: cwd, compile, outputFileRename })

    const localReferences = {}

    for (const targetPath of Object.keys(definitions)) {
        resolveTypescriptSchema({ definitions, targetPath, cwd, outputFiles, localReferences })
    }

    return compile ? outputFiles : outputFiles.filter((f) => f.type !== 'validator')
}

export async function generate({
    globs,
    ignore,
    extension,
    compile,
    outputFileRename,
    clean,
}: {
    globs: string[]
    ignore: string[]
    extension: string
    compile: boolean
    outputFileRename: (path: string) => string
    clean: boolean
}): Promise<void> {
    const cwd = process.cwd()
    const entries = await expandGlobs({ patterns: globs, ignore: [`!${extension}`, ...ignore], cwd, extension })

    const outputFiles = await compileOutputFiles(entries, {
        cwd,
        outputFileRename,
        compile,
    })

    if (clean) {
        for (const of of outputFiles) {
            if (of.clean !== undefined) {
                of.clean()
            }
        }
    }

    const prettier = maybeLoadPrettier()

    const data = outputFiles.reduce<Record<string, ReferenceData>>(
        (agg, x) => ({
            ...agg,
            ...x.data,
        }),
        {}
    )

    const duplicates = entriesOf(groupBy(entriesOf(data), ([, values]) => `${values.srcPath}${values.symbolName}`)).filter(
        ([, records]) => records.length > 1
    )
    for (const duplicateGroup of duplicates.map((d) => second(d))) {
        for (const [i, [key]] of enumerate(duplicateGroup)) {
            // data[key].referenceName = `${data[key].referenceName}${i}`
            data[key]!.symbolName = `${data[key]!.symbolName}${i}`
            data[key]!.uniqueSymbolName = `${data[key]!.uniqueSymbolName}${i}`
        }
    }

    const variablesMap = Object.fromEntries(
        Object.entries(data)
            .map(([key, vals]) => Object.entries(vals).map(([varName, value]) => [`${key}:${varName}`, value] as const))
            .flat()
    )

    for (const { targetPath, template, type, prettify } of outputFiles) {
        fs.mkdirSync(path.dirname(targetPath), { recursive: true })
        // allow nested references up to 3 deep
        let contents = template
        for (const _ of range(3)) {
            contents = renderTemplate(contents, variablesMap)
        }
        fs.writeFileSync(targetPath, prettify ? await formatFile(prettier, contents, targetPath, type) : contents)
    }
}
