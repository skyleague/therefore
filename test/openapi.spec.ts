import { compileOutputFiles } from '../src/commands/generate/generate.js'

import { expect, it } from 'vitest'

it('output generation', async () => {
    expect(
        await compileOutputFiles(['src/lib/primitives/restclient/openapi.schema.ts'], {
            outputFileRename: (file: string) => file.replace('.ts', '.type.ts'),
            cwd: process.cwd(),
            compile: true,
        })
    ).toMatchSnapshot()
})
