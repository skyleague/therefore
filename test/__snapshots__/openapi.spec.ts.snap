// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`output generation 1`] = `
[
  {
    "data": {
      "000109-000": {
        "referenceName": "{{000109-000:symbolName}}",
        "srcPath": "src/lib/primitives/restclient/openapi.schema.type.ts",
        "symbolName": "PathItem",
        "uniqueSymbolName": "{{000109-000:symbolName}}3178",
      },
      "000145-000": {
        "referenceName": "{{000145-000:symbolName}}",
        "srcPath": "src/lib/primitives/restclient/openapi.schema.type.ts",
        "symbolName": "Operation",
        "uniqueSymbolName": "{{000145-000:symbolName}}3178",
      },
      "000177-000": {
        "referenceName": "{{000177-000:symbolName}}",
        "srcPath": "src/lib/primitives/restclient/openapi.schema.type.ts",
        "symbolName": "Parameter",
        "uniqueSymbolName": "{{000177-000:symbolName}}3178",
      },
      "00018-000": {
        "referenceName": "{{00018-000:symbolName}}",
        "srcPath": "src/lib/primitives/restclient/openapi.schema.type.ts",
        "symbolName": "OpenapiV3",
        "uniqueSymbolName": "{{00018-000:symbolName}}",
      },
      "000270-000": {
        "referenceName": "{{000270-000:symbolName}}",
        "srcPath": "src/lib/primitives/restclient/openapi.schema.type.ts",
        "symbolName": "Schema",
        "uniqueSymbolName": "{{000270-000:symbolName}}3178",
      },
      "000272-000": {
        "referenceName": "{{000272-000:symbolName}}",
        "srcPath": "src/lib/primitives/restclient/openapi.schema.type.ts",
        "symbolName": "Reference",
        "uniqueSymbolName": "{{000272-000:symbolName}}3178",
      },
      "000277-000": {
        "referenceName": "{{000277-000:symbolName}}",
        "srcPath": "src/lib/primitives/restclient/openapi.schema.type.ts",
        "symbolName": "Discriminator",
        "uniqueSymbolName": "{{000277-000:symbolName}}3178",
      },
      "000289-000": {
        "referenceName": "{{000289-000:symbolName}}",
        "srcPath": "src/lib/primitives/restclient/openapi.schema.type.ts",
        "symbolName": "XML",
        "uniqueSymbolName": "{{000289-000:symbolName}}3178",
      },
      "00030-000": {
        "referenceName": "{{00030-000:symbolName}}",
        "srcPath": "src/lib/primitives/restclient/openapi.schema.type.ts",
        "symbolName": "Info",
        "uniqueSymbolName": "{{00030-000:symbolName}}3178",
      },
      "000305-000": {
        "referenceName": "{{000305-000:symbolName}}",
        "srcPath": "src/lib/primitives/restclient/openapi.schema.type.ts",
        "symbolName": "MediaType",
        "uniqueSymbolName": "{{000305-000:symbolName}}3178",
      },
      "000315-000": {
        "referenceName": "{{000315-000:symbolName}}",
        "srcPath": "src/lib/primitives/restclient/openapi.schema.type.ts",
        "symbolName": "Example",
        "uniqueSymbolName": "{{000315-000:symbolName}}3178",
      },
      "000329-000": {
        "referenceName": "{{000329-000:symbolName}}",
        "srcPath": "src/lib/primitives/restclient/openapi.schema.type.ts",
        "symbolName": "Encoding",
        "uniqueSymbolName": "{{000329-000:symbolName}}3178",
      },
      "000359-000": {
        "referenceName": "{{000359-000:symbolName}}",
        "srcPath": "src/lib/primitives/restclient/openapi.schema.type.ts",
        "symbolName": "Header",
        "uniqueSymbolName": "{{000359-000:symbolName}}3178",
      },
      "000367-000": {
        "referenceName": "{{000367-000:symbolName}}",
        "srcPath": "src/lib/primitives/restclient/openapi.schema.type.ts",
        "symbolName": "RequestBody",
        "uniqueSymbolName": "{{000367-000:symbolName}}3178",
      },
      "000376-000": {
        "referenceName": "{{000376-000:symbolName}}",
        "srcPath": "src/lib/primitives/restclient/openapi.schema.type.ts",
        "symbolName": "Responses",
        "uniqueSymbolName": "{{000376-000:symbolName}}3178",
      },
      "00038-000": {
        "referenceName": "{{00038-000:symbolName}}",
        "srcPath": "src/lib/primitives/restclient/openapi.schema.type.ts",
        "symbolName": "Contact",
        "uniqueSymbolName": "{{00038-000:symbolName}}3178",
      },
      "000392-000": {
        "referenceName": "{{000392-000:symbolName}}",
        "srcPath": "src/lib/primitives/restclient/openapi.schema.type.ts",
        "symbolName": "Response",
        "uniqueSymbolName": "{{000392-000:symbolName}}3178",
      },
      "000407-000": {
        "referenceName": "{{000407-000:symbolName}}",
        "srcPath": "src/lib/primitives/restclient/openapi.schema.type.ts",
        "symbolName": "Link",
        "uniqueSymbolName": "{{000407-000:symbolName}}3178",
      },
      "000410-000": {
        "referenceName": "{{000410-000:symbolName}}",
        "srcPath": "src/lib/primitives/restclient/openapi.schema.type.ts",
        "symbolName": "Callback",
        "uniqueSymbolName": "{{000410-000:symbolName}}3178",
      },
      "00043-000": {
        "referenceName": "{{00043-000:symbolName}}",
        "srcPath": "src/lib/primitives/restclient/openapi.schema.type.ts",
        "symbolName": "License",
        "uniqueSymbolName": "{{00043-000:symbolName}}3178",
      },
      "000457-000": {
        "referenceName": "{{000457-000:symbolName}}",
        "srcPath": "src/lib/primitives/restclient/openapi.schema.type.ts",
        "symbolName": "Components",
        "uniqueSymbolName": "{{000457-000:symbolName}}3178",
      },
      "000462-000": {
        "referenceName": "{{000462-000:symbolName}}",
        "srcPath": "src/lib/primitives/restclient/openapi.schema.type.ts",
        "symbolName": "SecurityScheme",
        "uniqueSymbolName": "{{000462-000:symbolName}}3178",
      },
      "000469-000": {
        "referenceName": "{{000469-000:symbolName}}",
        "srcPath": "src/lib/primitives/restclient/openapi.schema.type.ts",
        "symbolName": "APIKeySecurityScheme",
        "uniqueSymbolName": "{{000469-000:symbolName}}3178",
      },
      "000476-000": {
        "referenceName": "{{000476-000:symbolName}}",
        "srcPath": "src/lib/primitives/restclient/openapi.schema.type.ts",
        "symbolName": "HTTPSecurityScheme",
        "uniqueSymbolName": "{{000476-000:symbolName}}3178",
      },
      "00048-000": {
        "referenceName": "{{00048-000:symbolName}}",
        "srcPath": "src/lib/primitives/restclient/openapi.schema.type.ts",
        "symbolName": "ExternalDocumentation",
        "uniqueSymbolName": "{{00048-000:symbolName}}3178",
      },
      "000482-000": {
        "referenceName": "{{000482-000:symbolName}}",
        "srcPath": "src/lib/primitives/restclient/openapi.schema.type.ts",
        "symbolName": "OAuth2SecurityScheme",
        "uniqueSymbolName": "{{000482-000:symbolName}}3178",
      },
      "000492-000": {
        "referenceName": "{{000492-000:symbolName}}",
        "srcPath": "src/lib/primitives/restclient/openapi.schema.type.ts",
        "symbolName": "OAuthFlows",
        "uniqueSymbolName": "{{000492-000:symbolName}}3178",
      },
      "000499-000": {
        "referenceName": "{{000499-000:symbolName}}",
        "srcPath": "src/lib/primitives/restclient/openapi.schema.type.ts",
        "symbolName": "ImplicitOAuthFlow",
        "uniqueSymbolName": "{{000499-000:symbolName}}3178",
      },
      "000506-000": {
        "referenceName": "{{000506-000:symbolName}}",
        "srcPath": "src/lib/primitives/restclient/openapi.schema.type.ts",
        "symbolName": "PasswordOAuthFlow",
        "uniqueSymbolName": "{{000506-000:symbolName}}3178",
      },
      "000513-000": {
        "referenceName": "{{000513-000:symbolName}}",
        "srcPath": "src/lib/primitives/restclient/openapi.schema.type.ts",
        "symbolName": "ClientCredentialsFlow",
        "uniqueSymbolName": "{{000513-000:symbolName}}3178",
      },
      "000521-000": {
        "referenceName": "{{000521-000:symbolName}}",
        "srcPath": "src/lib/primitives/restclient/openapi.schema.type.ts",
        "symbolName": "AuthorizationCodeOAuthFlow",
        "uniqueSymbolName": "{{000521-000:symbolName}}3178",
      },
      "000527-000": {
        "referenceName": "{{000527-000:symbolName}}",
        "srcPath": "src/lib/primitives/restclient/openapi.schema.type.ts",
        "symbolName": "OpenIdConnectSecurityScheme",
        "uniqueSymbolName": "{{000527-000:symbolName}}3178",
      },
      "00056-000": {
        "referenceName": "{{00056-000:symbolName}}",
        "srcPath": "src/lib/primitives/restclient/openapi.schema.type.ts",
        "symbolName": "Server",
        "uniqueSymbolName": "{{00056-000:symbolName}}3178",
      },
      "00064-000": {
        "referenceName": "{{00064-000:symbolName}}",
        "srcPath": "src/lib/primitives/restclient/openapi.schema.type.ts",
        "symbolName": "ServerVariable",
        "uniqueSymbolName": "{{00064-000:symbolName}}3178",
      },
      "00067-000": {
        "referenceName": "{{00067-000:symbolName}}",
        "srcPath": "src/lib/primitives/restclient/openapi.schema.type.ts",
        "symbolName": "SecurityRequirement",
        "uniqueSymbolName": "{{00067-000:symbolName}}3178",
      },
      "00074-000": {
        "referenceName": "{{00074-000:symbolName}}",
        "srcPath": "src/lib/primitives/restclient/openapi.schema.type.ts",
        "symbolName": "Tag",
        "uniqueSymbolName": "{{00074-000:symbolName}}3178",
      },
      "00077-000": {
        "referenceName": "{{00077-000:symbolName}}",
        "srcPath": "src/lib/primitives/restclient/openapi.schema.type.ts",
        "symbolName": "Paths",
        "uniqueSymbolName": "{{00077-000:symbolName}}3178",
      },
    },
    "prettify": true,
    "relativeSource": "./openapi.schema",
    "targetPath": "src/lib/primitives/restclient/openapi.schema.type.ts",
    "template": "/**
* Generated by @skyleague/therefore@v1.0.0-local
* Do not manually touch this
*/
/* eslint-disable */


/**
 * The description of OpenAPI v3.0.x documents, as defined by https://spec.openapis.org/oas/v3.0.3
 */
export interface {{00018-000:symbolName}} {
    openapi: string
    info: {{00030-000:referenceName}}
    externalDocs?: {{00048-000:referenceName}}
    servers?: ({{00056-000:referenceName}})[]
    security?: ({{00067-000:referenceName}})[]
    tags?: ({{00074-000:referenceName}})[]
    paths: {{00077-000:referenceName}}
    components?: {{000457-000:referenceName}}
    [k: \`x-\${string}\`]: unknown
}

export interface {{00030-000:symbolName}} {
    title: string
    description?: string
    termsOfService?: string
    contact?: {{00038-000:referenceName}}
    license?: {{00043-000:referenceName}}
    version: string
    [k: \`x-\${string}\`]: unknown
}

export interface {{00038-000:symbolName}} {
    name?: string
    url?: string
    email?: string
    [k: \`x-\${string}\`]: unknown
}

export interface {{00043-000:symbolName}} {
    name: string
    url?: string
    [k: \`x-\${string}\`]: unknown
}

export interface {{00048-000:symbolName}} {
    description?: string
    url: string
    [k: \`x-\${string}\`]: unknown
}

export interface {{00056-000:symbolName}} {
    url: string
    description?: string
    variables?: {
        [k: string]: {{00064-000:referenceName}}
    }
    [k: \`x-\${string}\`]: unknown
}

export interface {{00064-000:symbolName}} {
    enum?: (string)[]
    default: string
    description?: string
    [k: \`x-\${string}\`]: unknown
}

export interface {{00067-000:symbolName}} {
    [k: string]: (string)[]
}

export interface {{00074-000:symbolName}} {
    name: string
    description?: string
    externalDocs?: {{00048-000:referenceName}}
    [k: \`x-\${string}\`]: unknown
}

export interface {{00077-000:symbolName}} {
    [k: \`/\${string}\`]: {{000109-000:referenceName}}
    [k: \`x-\${string}\`]: unknown
}

export interface {{000109-000:symbolName}} {
    get?: {{000145-000:referenceName}}
    put?: {{000145-000:referenceName}}
    post?: {{000145-000:referenceName}}
    delete?: {{000145-000:referenceName}}
    options?: {{000145-000:referenceName}}
    head?: {{000145-000:referenceName}}
    patch?: {{000145-000:referenceName}}
    trace?: {{000145-000:referenceName}}
    '$ref'?: string
    summary?: string
    description?: string
    servers?: ({{00056-000:referenceName}})[]
    parameters?: ({{000177-000:referenceName}} | {{000272-000:referenceName}})[]
    [k: \`x-\${string}\`]: unknown
}

export interface {{000145-000:symbolName}} {
    tags?: (string)[]
    summary?: string
    description?: string
    externalDocs?: {{00048-000:referenceName}}
    operationId?: string
    parameters?: ({{000177-000:referenceName}} | {{000272-000:referenceName}})[]
    requestBody?: {{000367-000:referenceName}} | {{000272-000:referenceName}}
    responses: {{000376-000:referenceName}}
    callbacks?: {
        [k: string]: {{000410-000:referenceName}} | {{000272-000:referenceName}}
    }
    /**
     * @default false
     */
    deprecated?: boolean
    security?: ({{00067-000:referenceName}})[]
    servers?: ({{00056-000:referenceName}})[]
    [k: \`x-\${string}\`]: unknown
}

export interface {{000177-000:symbolName}} {
    name: string
    in: string
    description?: string
    /**
     * @default false
     */
    required?: boolean
    /**
     * @default false
     */
    deprecated?: boolean
    /**
     * @default false
     */
    allowEmptyValue?: boolean
    style?: string
    explode?: boolean
    /**
     * @default false
     */
    allowReserved?: boolean
    schema?: {{000270-000:referenceName}} | {{000272-000:referenceName}}
    content?: {
        [k: string]: {{000305-000:referenceName}}
    }
    example?: unknown
    examples?: {
        [k: string]: {{000315-000:referenceName}} | {{000272-000:referenceName}}
    }
    [k: \`x-\${string}\`]: unknown
}

export interface {{000270-000:symbolName}} {
    title?: string
    multipleOf?: number
    maximum?: number
    /**
     * @default false
     */
    exclusiveMaximum?: boolean
    minimum?: number
    /**
     * @default false
     */
    exclusiveMinimum?: boolean
    maxLength?: number
    /**
     * @default 0
     */
    minLength?: number
    pattern?: string
    maxItems?: number
    /**
     * @default 0
     */
    minItems?: number
    /**
     * @default false
     */
    uniqueItems?: boolean
    maxProperties?: number
    /**
     * @default 0
     */
    minProperties?: number
    required?: [string,  ...(string)[]]
    enum?: [unknown,  ...(unknown)[]]
    type?: 'array' | 'boolean' | 'integer' | 'number' | 'object' | 'string'
    not?: {{000270-000:referenceName}} | {{000272-000:referenceName}}
    allOf?: ({{000270-000:referenceName}} | {{000272-000:referenceName}})[]
    oneOf?: ({{000270-000:referenceName}} | {{000272-000:referenceName}})[]
    anyOf?: ({{000270-000:referenceName}} | {{000272-000:referenceName}})[]
    items?: {{000270-000:referenceName}} | {{000272-000:referenceName}}
    properties?: {
        [k: string]: {{000270-000:referenceName}} | {{000272-000:referenceName}}
    }
    /**
     * @default true
     */
    additionalProperties?: {{000270-000:referenceName}} | {{000272-000:referenceName}} | boolean
    description?: string
    format?: string
    default?: unknown
    /**
     * @default false
     */
    nullable?: boolean
    discriminator?: {{000277-000:referenceName}}
    /**
     * @default false
     */
    readOnly?: boolean
    /**
     * @default false
     */
    writeOnly?: boolean
    example?: unknown
    externalDocs?: {{00048-000:referenceName}}
    /**
     * @default false
     */
    deprecated?: boolean
    xml?: {{000289-000:referenceName}}
    [k: \`x-\${string}\`]: unknown
}

export interface {{000272-000:symbolName}} {
    '$ref': string
}

export interface {{000277-000:symbolName}} {
    propertyName: string
    mapping?: {
        [k: string]: string
    }
}

export interface {{000289-000:symbolName}} {
    name?: string
    namespace?: string
    prefix?: string
    /**
     * @default false
     */
    attribute?: boolean
    /**
     * @default false
     */
    wrapped?: boolean
    [k: \`x-\${string}\`]: unknown
}

export interface {{000305-000:symbolName}} {
    schema?: {{000270-000:referenceName}} | {{000272-000:referenceName}}
    example?: unknown
    examples?: {
        [k: string]: {{000315-000:referenceName}} | {{000272-000:referenceName}}
    }
    encoding?: {
        [k: string]: {{000329-000:referenceName}}
    }
    [k: \`x-\${string}\`]: unknown
}

export interface {{000315-000:symbolName}} {
    summary?: string
    description?: string
    value?: unknown
    externalValue?: string
    [k: \`x-\${string}\`]: unknown
}

export interface {{000329-000:symbolName}} {
    contentType?: string
    headers?: {
        [k: string]: {{000359-000:referenceName}} | {{000272-000:referenceName}}
    }
    style?: 'form' | 'spaceDelimited' | 'pipeDelimited' | 'deepObject'
    explode?: boolean
    /**
     * @default false
     */
    allowReserved?: boolean
}

export interface {{000359-000:symbolName}} {
    description?: string
    /**
     * @default false
     */
    required?: boolean
    /**
     * @default false
     */
    deprecated?: boolean
    /**
     * @default false
     */
    allowEmptyValue?: boolean
    /**
     * @default 'simple'
     */
    style?: 'simple'
    explode?: boolean
    /**
     * @default false
     */
    allowReserved?: boolean
    schema?: {{000270-000:referenceName}} | {{000272-000:referenceName}}
    content?: {
        [k: string]: {{000305-000:referenceName}}
    }
    example?: unknown
    examples?: {
        [k: string]: {{000315-000:referenceName}} | {{000272-000:referenceName}}
    }
    [k: \`x-\${string}\`]: unknown
}

export interface {{000367-000:symbolName}} {
    description?: string
    content: {
        [k: string]: {{000305-000:referenceName}}
    }
    /**
     * @default false
     */
    required?: boolean
    [k: \`x-\${string}\`]: unknown
}

export interface {{000376-000:symbolName}} {
    default?: {{000392-000:referenceName}} | {{000272-000:referenceName}}
    [k: string]: {{000392-000:referenceName}} | {{000272-000:referenceName}} | unknown
    [k: \`x-\${string}\`]: unknown
}

export interface {{000392-000:symbolName}} {
    description: string
    headers?: {
        [k: string]: {{000359-000:referenceName}} | {{000272-000:referenceName}}
    }
    content?: {
        [k: string]: {{000305-000:referenceName}}
    }
    links?: {
        [k: string]: {{000407-000:referenceName}} | {{000272-000:referenceName}}
    }
    [k: \`x-\${string}\`]: unknown
}

export interface {{000407-000:symbolName}} {
    operationId?: string
    operationRef?: string
    parameters?: {
        [k: string]: unknown
    }
    requestBody?: unknown
    description?: string
    server?: {{00056-000:referenceName}}
    [k: \`x-\${string}\`]: unknown
}

export interface {{000410-000:symbolName}} {
    [k: string]: {{000109-000:referenceName}} | unknown
    [k: \`x-\${string}\`]: unknown
}

export interface {{000457-000:symbolName}} {
    schemas?: {
        [k: string]: {{000270-000:referenceName}} | {{000272-000:referenceName}}
    }
    responses?: {
        [k: string]: {{000272-000:referenceName}} | {{000392-000:referenceName}}
    }
    parameters?: {
        [k: string]: {{000272-000:referenceName}} | {{000177-000:referenceName}}
    }
    examples?: {
        [k: string]: {{000272-000:referenceName}} | {{000315-000:referenceName}}
    }
    requestBodies?: {
        [k: string]: {{000272-000:referenceName}} | {{000367-000:referenceName}}
    }
    headers?: {
        [k: string]: {{000272-000:referenceName}} | {{000359-000:referenceName}}
    }
    securitySchemes?: {
        [k: string]: {{000272-000:referenceName}} | {{000462-000:referenceName}}
    }
    links?: {
        [k: string]: {{000272-000:referenceName}} | {{000407-000:referenceName}}
    }
    callbacks?: {
        [k: string]: {{000272-000:referenceName}} | {{000410-000:referenceName}}
    }
    [k: \`x-\${string}\`]: unknown
}

export type {{000462-000:symbolName}} = {{000469-000:referenceName}} | {{000476-000:referenceName}} | {{000482-000:referenceName}} | {{000527-000:referenceName}}

export interface {{000469-000:symbolName}} {
    type: 'apiKey'
    name: string
    in: 'header' | 'query' | 'cookie'
    description?: string
    [k: \`x-\${string}\`]: unknown
}

export type {{000476-000:symbolName}} = {
    scheme?: string
} | {
    scheme?: unknown
}

export interface {{000482-000:symbolName}} {
    type: 'oauth2'
    flows: {{000492-000:referenceName}}
    description?: string
    [k: \`x-\${string}\`]: unknown
}

export interface {{000492-000:symbolName}} {
    implicit?: {{000499-000:referenceName}}
    password?: {{000506-000:referenceName}}
    clientCredentials?: {{000513-000:referenceName}}
    authorizationCode?: {{000521-000:referenceName}}
    [k: \`x-\${string}\`]: unknown
}

export interface {{000499-000:symbolName}} {
    authorizationUrl: string
    refreshUrl?: string
    scopes: {
        [k: string]: string
    }
    [k: \`x-\${string}\`]: unknown
}

export interface {{000506-000:symbolName}} {
    tokenUrl: string
    refreshUrl?: string
    scopes: {
        [k: string]: string
    }
    [k: \`x-\${string}\`]: unknown
}

export interface {{000513-000:symbolName}} {
    tokenUrl: string
    refreshUrl?: string
    scopes: {
        [k: string]: string
    }
    [k: \`x-\${string}\`]: unknown
}

export interface {{000521-000:symbolName}} {
    authorizationUrl: string
    tokenUrl: string
    refreshUrl?: string
    scopes: {
        [k: string]: string
    }
    [k: \`x-\${string}\`]: unknown
}

export interface {{000527-000:symbolName}} {
    type: 'openIdConnect'
    openIdConnectUrl: string
    description?: string
    [k: \`x-\${string}\`]: unknown
}

",
    "type": "typescript",
  },
]
`;
